include/unifex/task.hpp

```diff
@@ -74,112 +75,109 @@ template <typename T>
-          break;
-        default:
-          break;
-      }
+template <>
+struct _return_value_or_void<void> {
+  struct type {
+    void return_void() noexcept {
+      expected_.reset_value();
+      unifex::activate_union_member(expected_.value_);
+      expected_.state_ = _state::value;
     }
+    _expected<void> expected_;
+  };
+};
+
+template <typename T>
+struct _promise {
+  struct type : _promise_base, _return_value_or_void<T>::type {
+    using result_type = T;
 
     typename _task<T>::type get_return_object() noexcept {
       return typename _task<T>::type{
           coro::coroutine_handle<type>::from_promise(*this)};
     }
 
-    coro::suspend_always initial_suspend() noexcept {
-      return {};
-    }
-
     auto final_suspend() noexcept {
-      struct awaiter {
-        bool await_ready() noexcept {
-          return false;
-        }
+      struct awaiter : _final_suspend_awaiter_base {
         auto await_suspend(coro::coroutine_handle<type> h) noexcept {
           return h.promise().continuation_;
         }
-        void await_resume() noexcept {}
       };
       return awaiter{};
     }
 
-    coro::coroutine_handle<> unhandled_done() noexcept {
-      return doneCallback_(continuation_.address());
-    }
-
     void unhandled_exception() noexcept {
-      reset_value();
-      unifex::activate_union_member(exception_, std::current_exception());
-      state_ = state::exception;
+      this->expected_.reset_value();
+      unifex::activate_union_member(this->expected_.exception_, std::current_exception());
+      this->expected_.state_ = _state::exception;
     }
 
-    template(typename Value)
+    template(typename Value = T)
         (requires callable<decltype(unifex::await_transform), type&, Value>)
     auto await_transform(Value&& value)
         noexcept(is_nothrow_callable_v<decltype(unifex::await_transform), type&, Value>)
         -> callable_result_t<decltype(unifex::await_transform), type&, Value> {
       return unifex::await_transform(*this, (Value&&)value);
     }
 
-    template(typename Value)
-        (requires convertible_to<Value, T> AND constructible_from<T, Value>)
-    void return_value(Value&& value) noexcept(
-        std::is_nothrow_constructible_v<T, Value>) {
-      reset_value();
-      unifex::activate_union_member(value_, (Value &&) value);
-      state_ = state::value;
-    }
-
-    type() noexcept {}
-
-    ~type() {
-      reset_value();
-    }
-
     decltype(auto) result() {
-      if (state_ == state::exception) {
-        std::rethrow_exception(std::move(exception_).get());
-      }
-      return std::move(value_).get();
-    }
-
-    template <typename Func>
-    friend void
-    tag_invoke(tag_t<visit_continuations>, const type& p, Func&& func) {
-      if (p.info_) {
-        visit_continuations(*p.info_, (Func &&) func);
+      if (this->expected_.state_ == _state::exception) {
+        std::rethrow_exception(std::move(this->expected_.exception_).get());
```

Comment:
Does this need to be enclosed in `#ifdef` for `-fno-exceptions`?

Reply:
Eventually all the `rethrow_exception`s in the code will need to be replaced. The issue is, replaced with _what_? With `-fno-exceptions`, we can never reach this particular `rethrow_exception` until we add support for `co_await co_error(...)`, at which point we'll need to handle the error as we're doing for `unhandled_done`. That's a much bigger change, and should wait for a separate PR.

And until we replace _all_ the `rethrow_exception`s, replacing _this_ one -- when it can never get hit -- doesn't seem to buy us anything.

I'm inclined to leave it alone until we can do it right.